text/metrics/0238-pipeline-monitoring.md

```diff
@@ -0,0 +1,484 @@
+used in log messages to describe data that was tried at least once and
+will not be retried, which matches the processor's definition of
+`refused` in the sense that data was submitted to the next component
+in the pipeline and failed and does not match the processor's
+definition `dropped`.
+
+As the exporter helper is not part of a processor framework, it does
+not have a conventional way to count dropped items.  When the
+queue-sender is enabled and the queue is full, items are dropped in
+the standard sense, but they are counted using an `enqueue_failed`
+metric.
+
+## Proposed semantic conventions
+
+Following the analysis above, the main problem being addressed is
+confusion over the meaning of "dropped", which is sometimes success
+and sometimes failure.  The use of a single metric with optional
+attributes allows us to explicitly count success and failure while
+optionally counting additional dimensions.  As we will see, this
+allows introducing newly-distinct outcomes without breaking past
+conventions.
+
+For example, the term "rejected" has a formal definition in
+OpenTelemetry that is not expressed by existing metrics.  An item of
+telemetry is considered rejected when it is included in a successful
+request but was individually dropped (for stated reasons) and should
+not be retried; these items were successfully sent but dropped (due to
+partial success) after processing by the next stage in the pipeline.
+
+### Use of a single metric name
+
+The use of a single metric name is less confusing than the use of
+multiple metric names, because the user has to know only a single name
+to writing useful queries.  Users working with existing collector and
+SDK pipeline monitoring metrics have to remember at least three metric
+names and explicitly join them via custom metric queries.  For
+example, to calculate loss rate for an SDK using traditional pipeline
+metrics,
+
+```
+LossRate_MultipleMetrics = (dropped + failed) / (dropped + failed + success)
+```
+
+On the other hand, with a uniform boolean attribute indicating success
+or failure the resulting query is simpler.
+
+```
+LossRate_SingleMetric = items{success=false} / items{success=*}
+```
+
+In a typical metric query engine, after the user has entered the one
+metric name, attribute values will be automatically surfaced in the
+user interface, allowing them to make sense of the data and
+interactively build useful queries.  On the other hand, the user who
+has to query multiple metrics has to enter each metric name
+explicitly without help from the user interface.
+
+The proposed metric instrument would be named distinctly depending on
+whether it is a collector or an SDK, to prevent accidental aggregation
+of these timeseries.  The specified counter names would be:
+
+- `otelsdk.producer.items`: count of successful and failed items of
+  telemetry produced, by signal type, by an OpenTelemetry SDK.
+- `otelcol.receiver.items`: count of successful and failed items of
```

Comment:
If I have a connector configured, I assume I get both the `otelcol.exporter.items` and `otelcol.receiver.items` metrics emitted for the connector, right?

Let's say I configured the [Count connector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.92.0/connector/countconnector/README.md) on a traces pipeline, as described in the example in the component's README. The count connector then accepts traces on the `traces/in` pipeline and creates metrics on the `metrics/out` pipeline.

I imagine the `otelcol.exporter.items` metric for the count connector would count the incoming spans on the trace pipeline. What would be the `otel.outcome` for those correctly consumed spans? Would it be `consumed` or rather `unsampled`? These logs aren't shipped anywhere by the component, they are "swallowed" by the connector if I understand correctly.

I imagine the `otelcol.receiver.items` metric for the count connector would count the metrics created on the metrics pipeline, with the `otel.outcome` set to `consumed`.

Reply:
Is the consume operation synchronous? I think the traces/in pipeline will wait until the metrics/out pipeline finishes the consume operation, so the outcome for traces/in will depend on the outcome for metrics/out. If metrics/out fails w/ a retryable status, maybe the producer will retry.

Since the count connector can produce more or fewer metric data points than arriving spans, I do not expect the item counts to match between the exporter and receiver, but I think the outcomes could match for synchronous operations. If the operation is asynchronous, the rules discussed in this proposal would apply -- the traces/in might see `consumed` while the metrics/out sees some sort of failure.

I don't see any problems, per se, just that the monitoring equations for connectors don't apply. I can't assume that the items_in == items_dropped + items_out.